File: src/main/java/dev/sky_lock/pocketlifevehicle/vehicle/CarEntities.java
package dev.sky_lock.pocketlifevehicle.vehicle;

import dev.sky_lock.pocketlifevehicle.PLVehicle;
import dev.sky_lock.pocketlifevehicle.item.ItemStackBuilder;
import dev.sky_lock.pocketlifevehicle.packet.ActionBar;
import dev.sky_lock.pocketlifevehicle.util.Formats;
import dev.sky_lock.pocketlifevehicle.util.Profiles;
import dev.sky_lock.pocketlifevehicle.vehicle.model.Capacity;
import dev.sky_lock.pocketlifevehicle.vehicle.model.Model;
import dev.sky_lock.pocketlifevehicle.vehicle.model.ModelList;
import org.bukkit.*;
import org.bukkit.entity.Item;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.tags.ItemTagType;
import org.bukkit.metadata.FixedMetadataValue;

import java.io.IOException;
import java.util.*;
import java.util.logging.Logger;
import java.util.stream.Collectors;

/**
 * @author sky_lock
 */

public class CarEntities {
    private static final Logger logger = PLVehicle.getInstance().getLogger();
    private static final Map<UUID, Car> entities = new HashMap<>();

    public static boolean spawn(UUID player, Model model, Location location, float fuel) {
        if (player == null || model == null || location == null) {
            return false;
        }
        if (location.getBlock().getType() != Material.AIR) {
            ActionBar.sendPacket(Bukkit.getPlayer(player), ChatColor.RED + "ブロックがあるので車を設置できません");
            return false;
        }

        Car car = null;
        if (model.getCapacity() == Capacity.ONE_SEAT) {
            car = new OneSeatCar(model);
        } else if (model.getCapacity() == Capacity.TWO_SEATS) {
            car = new TwoSeatsCar(model);
        } else if (model.getCapacity() == Capacity.FOR_SEATS) {
            car = new FourSeatsCar(model);
        }
        if (car == null) {
            return false;
        }
        car.getStatus().setFuel(fuel);
        car.spawn(location);
        kill(player);

        entities.put(player, car);
        return true;
    }

    public static boolean spawn(Car car) {
        return getOwner(car).map(owner -> spawn(owner, car.getModel(), car.getLocation(), car.getStatus().getFuel())).orElse(false);
    }

    public static void kill(UUID owner) {
        if (entities.containsKey(owner)) {
            Car car = entities.remove(owner);
            car.kill();
        }
    }

    public static void kill(Car car) {
        if (entities.containsValue(car)) {
            entities.values().remove(car);
            car.kill();
        }
    }

    private static void killAll() {
        entities.values().forEach(Car::kill);
    }

    public static void tow(Car car) {
        getOwner(car).ifPresent(owner -> tow(owner, car));
    }

    public static void tow(UUID owner) {
        Optional.ofNullable(entities.get(owner)).ifPresent(car -> tow(owner, car));
    }

    private static void tow(UUID owner, Car car) {
        Model model = car.getModel();
        ItemStack itemStack = ItemStackBuilder.of(model.getItemStack())
                .tag(PLVehicle.getInstance().createKey("owner"), ItemTagType.STRING, owner.toString())
                .lore("所有者: " + Profiles.getName(owner), "残燃料: " + Formats.truncateToOneDecimalPlace(car.getStatus().getFuel()))
                .itemFlags(ItemFlag.values())
                .build();
        Location location = car.getLocation();
        Item item = location.getWorld().dropItem(car.getLocation(), itemStack);
        item.setMetadata("mocar-fuel", new FixedMetadataValue(PLVehicle.getInstance(), car.getStatus().getFuel()));

        location.getWorld().playSound(location, Sound.BLOCK_IRON_DOOR_OPEN, 1F, 0.2F);
        kill(owner);
    }

    public static Car getCar(SeatArmorStand seat) {
        return entities.values().stream().filter(car -> car.contains(seat)).findFirst().orElse(null);
    }

    public static Car getCar(CarArmorStand basis) {
        return entities.values().stream().filter(car -> car.contains(basis)).findFirst().orElse(null);
    }

    static Set<CarEntity> getCarEntities() {
        return entities.entrySet().stream().map(entry -> new CarEntity(entry.getKey().toString(), entry.getValue().getModel().getId(), entry.getValue().getLocation(), entry.getValue().getStatus().getFuel())).collect(Collectors.toSet());
    }

    public static Car of(UUID player) {
        return entities.get(player);
    }

    public static Optional<UUID> getOwner(Car car) {
        return entities.entrySet().stream().filter(entry -> entry.getValue().equals(car)).findFirst().map(Map.Entry::getKey);
    }

    public static void spawnAll() {
        try {
            PLVehicle.getInstance().getCarStoreFile().load().forEach(carEntity -> {
                ModelList.of(carEntity.getModelId()).ifPresent(model -> {
                    CarEntities.spawn(carEntity.getOwner(), model, carEntity.getLocation(), carEntity.getFuel());
                });
            });
        } catch (IOException ex) {
            logger.warning("CarEntityの読み込みに失敗しました");
        }
    }

    public static void saveAll() {
        try {
            PLVehicle.getInstance().getCarStoreFile().save(CarEntities.getCarEntities());
            CarEntities.killAll();
        } catch (IOException ex) {
            logger.warning("CarEntityの保存に失敗しました");
        }
    }
}

File: src/main/kotlin/dev/sky_lock/pocketlifevehicle/vehicle/CarEntities.kt
package dev.sky_lock.pocketlifevehicle.vehicle

import dev.sky_lock.pocketlifevehicle.PLVehicle
import dev.sky_lock.pocketlifevehicle.item.ItemStackBuilder.Companion.of
import dev.sky_lock.pocketlifevehicle.util.Formats
import dev.sky_lock.pocketlifevehicle.util.Profiles
import dev.sky_lock.pocketlifevehicle.vehicle.model.Capacity
import dev.sky_lock.pocketlifevehicle.vehicle.model.Model
import org.bukkit.*
import org.bukkit.inventory.ItemFlag
import org.bukkit.persistence.PersistentDataType
import java.io.IOException
import java.util.*
import java.util.function.Consumer

/**
 * @author sky_lock
 */
object CarEntities {
    private val logger = PLVehicle.instance.logger
    private val entities: MutableMap<UUID, Car> = HashMap()

    fun spawn(player: UUID, model: Model, location: Location, fuel: Float): Boolean {
        if (location.block.type != Material.AIR) {
            Bukkit.getPlayer(player)!!.sendActionBar(ChatColor.RED.toString() + "ブロックがあるので車を設置できません")
            return false
        }
        var car: Car? = null
        val capacity = model.spec.capacity
        when {
            capacity === Capacity.ONE_SEAT -> {
                car = OneSeatCar(model)
            }
            capacity === Capacity.TWO_SEATS -> {
                car = TwoSeatsCar(model)
            }
            capacity === Capacity.FOR_SEATS -> {
                car = FourSeatsCar(model)
            }
        }
        if (car == null) {
            return false
        }
        car.status.fuel = fuel
        car.spawn(location)
        kill(player)
        entities[player] = car
        return true
    }

    fun spawn(car: Car): Boolean {
        val owner = getOwner(car) ?: return false
        return spawn(owner, car.model, car.location, car.status.fuel)
    }

    fun kill(owner: UUID) {
        if (entities.containsKey(owner)) {
            val car = entities.remove(owner)
            car!!.kill()
        }
    }

    fun kill(car: Car) {
        if (entities.containsValue(car)) {
            entities.values.remove(car)
            car.kill()
        }
    }

    private fun killAll() {
        entities.values.forEach(Consumer { obj: Car -> obj.kill() })
    }

    fun tow(car: Car) {
        getOwner(car)?.let { owner -> tow(owner, car) }
    }

    fun tow(owner: UUID) {
        val car = entities[owner] ?: return
        tow(owner, car)
    }

    private fun tow(owner: UUID, car: Car) {
        val model = car.model
        val itemStack = of(model.itemStack)
                .persistentData(PLVehicle.instance.createKey("owner"), PersistentDataType.STRING, owner.toString())
                .lore("所有者: " + Profiles.getName(owner), "残燃料: " + Formats.truncateToOneDecimalPlace(car.status.fuel))
                .itemFlags(*ItemFlag.values())
                .build()
        val location = car.location
        location.world.dropItem(car.location, itemStack)
        // item.setMetadata("mocar-fuel", new FixedMetadataValue(PLVehicle.getInstance(), car.getStatus().getFuel()));
        location.world.playSound(location, Sound.BLOCK_IRON_DOOR_OPEN, 1f, 0.2f)
        kill(owner)
    }

    fun getCar(seat: SeatArmorStand): Car? {
        return entities.values.find { car: Car -> car.contains(seat) }
    }

    fun getCar(basis: CarArmorStand): Car? {
        return entities.values.find { car: Car -> car.contains(basis) }
    }

    private val carEntities: Set<CarEntity>
        get() = entities.entries.map { entry -> CarEntity(entry.key.toString(), entry.value.model.id, entry.value.location, entry.value.status.fuel) }.toSet()

    fun of(player: UUID): Car? {
        return entities[player]
    }

    fun getOwner(car: Car): UUID? {
        return entities.entries.find { entry: Map.Entry<UUID, Car> -> entry.value == car }?.key
    }

    fun spawnAll() {
        try {
            PLVehicle.instance.entityStoreFile.load().forEach(Consumer { carEntity: CarEntity ->
                val model = Storage.MODEL.findById(carEntity.modelId)
                if (model != null) {
                    spawn(carEntity.owner, model, carEntity.location, carEntity.fuel)
                }
            })
        } catch (ex: IOException) {
            logger.warning("CarEntityの読み込みに失敗しました")
        }
    }

    fun saveAll() {
        try {
            PLVehicle.instance.entityStoreFile.save(carEntities)
            killAll()
        } catch (ex: IOException) {
            logger.warning("CarEntityの保存に失敗しました")
        }
    }
}